### PR TITLE
Add tests to check that requests are retried when writing/reading blobs on S3

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds tests to verify the behavior of the `S3BlobContainer` and its underlying AWS SDK client when the remote S3 service is responding errors or not responding at all. The expected behavior is that requests are retried multiple times before the client gives up and the `S3BlobContainer` bubbles up an exception.

The test verifies the behavior of `BlobContainer.writeBlob()` and `BlobContainer.readBlob()`. In the case of S3 writing a blob can be executed as a single upload or using multipart requests; the test checks both scenario by writing a small then a large blob.

### CODE REVIEW

**Comment by tlrx** [comment]
@original-brownbear Sorry for the time it took me to create these tests. I tried to reuse the `AmazonS3Fixture` but this one has evolved in a freaking beast that is not easily reusable in the context of this PR. The broken pipe/connection drop use case is not tested but I'd be happy to have some feedback from you already.

**Comment by original-brownbear** [comment]
@tlrx no worries I see an in-JVM http server so I like it already :) => will look into this properly very shortly :)

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (original-brownbear) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: I love this one :D so much more than the fixture code, maybe we could add a TODO to dry things up against the fixture here at some point? (then we could also port some tests that use the weird S3 mock right now to just run on top of this?)

**Review by original-brownbear — COMMENTED** [review]
Looks really nice except for the timeout situation, I added a suggestion for that though (hope it makes sense) :)

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (tlrx) [reply plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: It makes sense. I pushed 1d32f317f6f729576926d430c6b47b3c289e4030, let me know what you think

**Review by original-brownbear — APPROVED** [review]
LGTM, thanks @tlrx !

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (tlrx) [reply plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: Thanks!

The actual S3 fixture is much more complicated and has strangely evolved, but note that we are not exactly testing the same thing here. In this test, we only want the HTTP server to reply responses that serve our interest in testing the retries/timeout logic. The AmazonS3Fixture on his side exist to allow to test against a real mocked S3 service. The responses are them much more elaborated. Or maybe I'm missing something in your suggestion?

**Review by original-brownbear — APPROVED** [review]
LGTM2 :D

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (original-brownbear) [reply plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: Yea true, I got a little carried away with my wish to move the fixture into the JUnit JVM :D Don't worry about it, I'll open a PR with a suggestion on this, but you're right the fixture is pretty complicated so it won't be so quick to move it to the same http server infra ... forget I said anything for now :)

Comment (tlrx) [reply plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: There's definitely a room for improvement there, let me know if you want some help for code or review :)

**Comment by original-brownbear** [comment]
@tlrx before I forget, why not merge this to `7.4` as well?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1
Jenkins run elasticsearch-ci/bwc

(unrelated failures)

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (ywelsch) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: we're relying on the fact that the implementation is ignoring the read limit, as the AWS SDK sets a low limit here (~131kb).

The mark/reset behavior of the SDK is also dependent on whether we have a `BufferedInputStream` (as you pointed out).
It would be good to test the behavior of the actual inputstream that we are passing in.

In particular, the assertion added by #45224 is not sufficient to guarantee us retry behavior.

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
Thanks @original-brownbear and @ywelsch !

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java
Comment (tlrx) [reply plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/S3BlobContainerRetriesTests.java]: > we're relying on the fact that the implementation is ignoring the read limit, as the AWS SDK sets a low limit here (~131kb).

As far as I understand the InputStream documentation, it seems that the read limit might or might not be respected by implementations. In any case the AWS SDK provides a default read limit (131kb) when marking the input stream but does not really expect the stream to respect the limit. The only exception is when the provided input stream is a `BufferedInputStream` and in this case the request client options read limit must be configured accordingly.

I've been confused by this because at first I tried to use a `BufferedInputStream` in the test class but the AWS SDK has some special treatment for such streams, and since we don't provide buffered input stream to the client when snapshotting data I think it just does not make sense to use it here.

 I opened #45847 to beef up the tests a bit so that we sometimes upload blobs that are large than the default read limit and that sometimes we consumes part of the request body. 

>  It would be good to test the behavior of the actual inputstream that we are passing in.

I agree, I'll add such tests as a follow up.

**Pull request merged**
